--=======================================================================================================>
-- Important note: Precision checks currently only for 'players' and the 'localplayer', not 'parts'.

-- EnumName, EnumValue, AdditionalProperty
return {
	--======================================================================>
	{"PostSimulation", 1}, 
	{"PreSimulation", 2}, 
	{"PreRender", 3}, 
	--======================================================================>
}

--=======================================================================================================>